### DISCUSSION TITLE
What is the process understanding of this part of the code？

### DISCUSSION THREAD

**Discussion by dannyfe** [juicedata/juicefs#1193] [Q&A] [answered]
How to write after hitting a slice or creating a new slice is not well understood. Can someone help you understand it

https://github.com/juicedata/juicefs/blob/f07634f03c2ac50ec7683a61cbfab585b914f8de/pkg/chunk/cached_store.go#L337

func (c *wChunk) WriteAt(p []byte, off int64) (n int, err error) {
   if int(off)+len(p) > chunkSize {
      return 0, fmt.Errorf("write out of chunk boudary: %d > %d", int(off)+len(p), chunkSize)
   }
   if off < int64(c.uploaded) {
      return 0, fmt.Errorf("Cannot overwrite uploaded block: %d < %d", off, c.uploaded)
   }

   // Fill previous blocks with zeros
   if c.length < int(off) {
      zeros := make([]byte, int(off)-c.length)
      _, _ = c.WriteAt(zeros, int64(c.length)) //填充本次写入数据到上一次数据记录点得空白块
   }

   for n < len(p) {
      //off = off - s.off 当前slice相对于本次写入偏移量得差值 13M
      indx := c.index(int(off) + n)// (int(off) + n / c.store.conf.BlockSize) // 3
      boff := (int(off) + n) % c.store.conf.BlockSize // 1

      var bs = pageSize
      if indx > 0 || bs > c.store.conf.BlockSize {
         bs = c.store.conf.BlockSize
      }

      bi := boff / bs //16
      bo := boff % bs //0
      var page *Page
      if bi < len(c.pages[indx]) { //16 < 0
         page = c.pages[indx][bi]
      } else {
         page = allocPage(bs)
         page.Data = page.Data[:0]
         c.pages[indx] = append(c.pages[indx], page)
      }
      left := len(p) - n
      if bo+left > bs {
         page.Data = page.Data[:bs]
      } else if len(page.Data) < bo+left {
         page.Data = page.Data[:bo+left]
      }
      n += copy(page.Data[bo:], p[n:])
   }
   if int(off)+n > c.length {
      c.length = int(off) + n
   }
   return n, nil
}

**Comment by SandyXSD** [comment, marked as answer]
Basically this function is used to write data `p` to the slice (aka, wChunk). There is a field `pages` in wChunk that is used to cache data, so what the function does is just splitting `p` in an aligned way and copy bytes to `pages`.

As wChunk is split to several Blocks for further writing to the object storage, `pages` is a list of memory blocks. For better memory management, the first block in `pages` is split again to 64 KiB pages, in case `len(p)` is far smaller than BlockSize (defaults 4 MiB).

The following figure may help you better understand the process:
```
         64K 64K ...
wChunk: |_  _  _  _  ...  _|________|________|__...
               4 M             4M       4M      ...

     p:         |_______________________|
              offset               
```

**Reply by dannyfe** [reply]
I sorted out the process according to the reply, and the data writing process is much clearer. Thank you for your answer
